rasterio/_warp.pyx

```diff
@@ -585,3 +488,83 @@ def _calculate_default_transform(
     dst_height = nlines
 
     return dst_affine, dst_width, dst_height
+
+cdef void *_create_MEM_dataset(data, transform, crs, desc, src_count=None) except *:
```

Comment:
@dmwelch we also have [InMemoryRaster](https://github.com/mapbox/rasterio/blob/master/rasterio/_io.pyx#L1899) that is intended to handle the IO between numpy arrays and GDAL in memory datasets.  Was there something lacking in that implementation?

Usage of that class isn't particularly obvious or well documented yet.